feat(react/vue-searchbox): add suggestion type query support

mohdashraf010897 opened the pull request:
**PR Type** Feature Integration

**Description** Integration of new type of query `suggestion` to fetch suggestions of different types using a single API call.

- [x]  When we're making the suggestion query (whenever autosuggest is set to `true`, use the new suggestion type instead of the current manual calls for recent/popular suggestions.
- [x]  Remove manual API calls to `popular` suggestions and `recent` suggestions
- [x]  Update the popular/recent suggestions extraction logic from the suggestion API response
- [x] You don't need to display suggestions based on the value. The API would handle it. Currently, we display suggestions based on the value. For e.g, display recent suggestions when value is empty or popular and index suggestions when value is present.
- [x] Remove all the old prop names for supporting popular and recent suggestions.
- [x] Update docs.

[🎥  Loom Demo - 12-OCT-2021-02:54-AM](https://www.loom.com/share/8bfccff9100747e8a4fcf49c609b4177)
[🎥  Loom Demo - 13-OCT-2021-04:30-PM](https://www.loom.com/share/5585d09e362841cb908c6683e27607ab)

[📄  Docs PR](https://github.com/appbaseio/Docs/pull/241) 

[📔  Notion ](https://www.notion.so/appbase/FE-Libraries-Integration-for-suggestion-type-cba01831fea4460d9639dd3df41043b0)

mohdashraf010897 committed:
feat: add suggestion type query support for searchbox component

mohdashraf010897 committed:
feat(vue): add suggestion type query support for searchbox component

bietkul reviewed (changes requested):
@mohdashraf010897 
Functionality issues
1. Ranking is managed by BE so FE don't need any logic to display popular suggestions at top or bottom. Just render the list and change icon as per the type of suggestion.
![alt de](https://i.imgur.com/e1m64Ep.png)
2. The no. of popular suggestions high because of BE issue.
![alt w](https://i.imgur.com/XSbWvpW.jpg)

bietkul commented on packages/vue-searchbox/src/utils/helper.js:
```diff
@@ -226,3 +226,18 @@ export function extractModifierKeysFromFocusShortcuts(focusShortcutsArray) {
+export const suggestionTypes = {
+	Popular: 'popular',
+	Index: 'index',
+	Recent: 'recent'
```
Add one more type as `Promoted: 'promoted'`

bietkul commented on packages/vue-searchbox/src/components/SearchComponent.jsx:
```diff
@@ -190,6 +201,12 @@ const SearchComponent = {
 			clearOnQueryChange,
 			distinctField,
 			distinctFieldConfig,
+			enableRecentSuggestions,
```
@mohdashraf010897 The property name in searhcbox lib is `enableRecentSearches`. We don't want to break the existing functionality. So support both names. If `enableRecentSuggestions` is not defined then fallback to `enableRecentSearches`.

bietkul commented on packages/vue-searchbox/src/components/SearchComponent.jsx:
```diff
@@ -61,7 +61,12 @@ const SearchComponent = {
+  		popularSuggestionsConfig: VueTypes.object,
+ 		maxPredictedWords: VueTypes.number,
+ 		urlField: VueTypes.string,
+  		rankFeature: VueTypes.object,
```
You have missed some props. For e.g `enablePopularSuggestions`

mohdashraf010897 replied:
it is already there @bietkul on Line 18-19

bietkul commented on packages/vue-searchbox/src/components/SearchBox.jsx:
```diff
@@ -912,7 +899,8 @@ const SearchBoxWrapper = {
 				value=""
-				triggerQueryOnInit={!!context.props.enablePopularSuggestions}
+				type={queryTypes.Suggestion}
+				triggerQueryOnInit={!!context.props.autosuggest}
```
We only want to trigger query initially when `recentSuggestions` are enabled.

bietkul reviewed (changes requested):
Remove suggestions parsing logic from searchcomponent. The `suggestions` property is no longer needed

mohdashraf010897 committed:
fix: requested changes

removed renderpopularsuggestions slot/ prop - changed logic for rendering suggestions

mohdashraf010897 committed:
fix: triggerQueryOnInit logic now based on recentsearches

mohdashraf010897 committed:
feat: added start icon for promoted search type

bietkul committed:
build: v1.6.0-alpha

mohdashraf010897 committed:
feat(react-searchbox): added support for urlField, promoted results, categoryfield

mohdashraf010897 committed:
feat(vue-searchbox): added support for urlField, promoted results, categoryfield

mohdashraf010897 committed:
Merge branch 'feat/suggestion-type-integration' of https://github.com/appbaseio/searchbox into feat/suggestion-type-integration

mohdashraf010897 committed:
fix(react-searchbox): recentsuggestions query firing

mohdashraf010897 committed:
fix(react-vue): fixed css issue with icons and text indentation in dropdow

mohdashraf010897 committed:
final fix: addition of categoryField, urlField and promoted result support

mohdashraf010897 committed:
fix: var name for suggestionslist

mohdashraf010897 committed:
feat: added support for applystopwords and stopwords

mohdashraf010897 committed:
feat: added support for applystopwords and stopwords

mohdashraf010897 committed:
fix: icon alignment issue for promoted icon

mohdashraf010897 committed:
fix: Value doesn't get cleared on clearing search for next query,

mohdashraf010897 committed:
fix: highlight issue for predictive suggestions

mohdashraf010897 committed:
refactor: revert example file changes

mohdashraf010897 committed:
chore: upgrade packages

mohdashraf010897 committed:
fix: resolved conflicts

mohdashraf010897 committed:
fix: resolved conflicts
